apps/server/src/modules/oauth/service/oauth.service.ts

```diff
@@ -1,37 +1,98 @@
 import { DefaultEncryptionService, IEncryptionService } from '@shared/infra/encryption';
 import { Logger } from '@src/core/logger';
-import { FeathersJwtProvider } from '@src/modules/authorization';
 import { UserService } from '@src/modules/user';
-import { AxiosResponse } from 'axios';
 import jwt, { JwtPayload } from 'jsonwebtoken';
-import JwksRsa from 'jwks-rsa';
-import QueryString from 'qs';
-import { lastValueFrom, Observable } from 'rxjs';
-import { AuthorizationParams, OauthTokenResponse, TokenRequestPayload } from '../controller/dto';
+
+import { Configuration } from '@hpi-schul-cloud/commons';
+import { SystemDto } from '@src/modules/system/service/dto/system.dto';
+
+import { ProvisioningDto, ProvisioningService } from '@src/modules/provisioning';
+import { OauthTokenResponse, AuthorizationParams, TokenRequestPayload } from '@src/modules/oauth/controller/dto';
+
+import { TokenRequestMapper } from '../mapper/token-request.mapper';
+
 import { OAuthSSOError } from '../error/oauth-sso.error';
 import { IJwt } from '../interface/jwt.base.interface';
-import { TokenRequestMapper } from '../mapper/token-request.mapper';
+
 import { OAuthProcessDto } from './dto/oauth-process.dto';
+import { SystemService } from '../../system';
+import { OauthDataDto } from '../../provisioning/dto';
+import { UserMigrationService } from '../../user-migration';
+import { OauthAdapterService } from './oauth-adapter.service';
 
 @Injectable()
 export class OAuthService {
 	constructor(
 		private readonly userService: UserService,
-		private readonly jwtService: FeathersJwtProvider,
-		private readonly httpService: HttpService,
+		private readonly oauthAdapterService: OauthAdapterService,
 		@Inject(DefaultEncryptionService) private readonly oAuthEncryptionService: IEncryptionService,
-		private readonly logger: Logger
+		private readonly logger: Logger,
+		private readonly provisioningService: ProvisioningService,
+		private readonly systemService: SystemService,
+		private readonly userMigrationService: UserMigrationService
 	) {
 		this.logger.setContext(OAuthService.name);
 	}
 
+	async authenticateUser(authCode: string, systemId: string): Promise<{ user?: UserDO; redirect: string }> {
+		const system = await this.systemService.findOAuthById(systemId);
+		if (!system.id) {
+			// unreachable. System loaded from DB always has an ID
+			throw new UnauthorizedException(`System with id "${systemId}" does not exist.`);
+		}
+
+		const oauthConfig: OauthConfig = this.extractOauthConfigFromSystem(system);
+
+		const queryToken: OauthTokenResponse = await this.requestToken(authCode, oauthConfig);
+
+		await this.validateToken(queryToken.id_token, oauthConfig);
+
+		const data: OauthDataDto = await this.provisioningService.getData(
+			queryToken.access_token,
+			queryToken.id_token,
+			system.id
+		);
+
+		let redirect = this.getRedirectUrl(oauthConfig.provider, queryToken.id_token, oauthConfig.logoutEndpoint);
```

Comment:
the response is a combination of a user and where to redirect it. wheatear this information gets used to build a token or redirect the user directly is up to the calling UC/Service. Building it here also removes logic from the OAuth UC itself. However it would be nicer to have this all in Client as we discussed. We can plan this in the ne